src/main/java/dev/openfeature/sdk/ImmutableContext.java

```diff
@@ -86,10 +87,34 @@ public EvaluationContext merge(EvaluationContext overridingContext) {
         if (overridingContext.getTargetingKey() != null && !overridingContext.getTargetingKey().trim().equals("")) {
             newTargetingKey = overridingContext.getTargetingKey();
         }
-        Map<String, Value> merged = new HashMap<>();
 
-        merged.putAll(this.asMap());
-        merged.putAll(overridingContext.asMap());
+        Map<String, Value> merged = this.merge(this.asMap(), overridingContext.asMap());
+
         return new ImmutableContext(newTargetingKey, merged);
     }
+
+    /**
+     * Recursively merges the base map from this EvaluationContext with the passed EvaluationContext.
+     * 
+     * @param base base map to merge
+     * @param overriding overriding map to merge
+     * @return resulting merged map
+     */
+    private Map<String, Value> merge(Map<String, Value> base, Map<String, Value> overriding) {
+        Map<String, Value> merged = new HashMap<>();
+
+        merged.putAll(base);
+        for (Entry<String, Value> overridingEntry : overriding.entrySet()) {
+            String key = overridingEntry.getKey();
+            if (overridingEntry.getValue().isStructure() && merged.containsKey(key) && merged.get(key).isStructure()) {
+                Structure mergedValue = merged.get(key).asStructure();
+                Structure overridingValue = overridingEntry.getValue().asStructure();
+                Map<String, Value> newMap = this.merge(mergedValue.asMap(), overridingValue.asMap());
+                merged.put(key, new Value(new ImmutableContext(newMap)));
```

Comment:
I think we want a new `Structure` here, not a new `Context`.

Reply:
Changed to `ImmutableStructure`